src/environments.jl

```diff
@@ -4,79 +4,238 @@ struct Environments
+    for sweep in 1:2
+        order      = isodd(sweep) ? (1:Ny) : reverse(1:Ny)
+        Envs_above = constructEnvsAbove(A, guess, prev_mps, ops, col, sweep)
+        Env_below  = is_cu ? cuITensor(1.0) : ITensor(1.0)
+        @inbounds for row in order
+            # construct environment for the row, have to do this every time
+            @timeit "build row Env" begin
+                Env  = copy(Env_below)
+                tmp  = A[row, col]
+                tmp *= ops[row]
+                tmp *= prev_mps[row]
+                Env *= tmp
+                Env *= dag(prime(A[row, col]))
+                Env_below  = copy(Env)
+                next_row = isodd(sweep) ? row+1 : row -1
+                Env *= next_row > 0 ? copy(Envs_above[next_row]) : dummy 
+                #guess[row] = copy(Env)
+                #Env_below  *= guess[row] 
+            end
+            # get l-inds for svd
+            @timeit "do SVD" begin
+                if isodd(sweep) && row < Ny
+                    svd_linds     = setdiff(inds(guess[row]),IndexSet(commonind(guess[row], guess[row+1])))
+                    tsvd          = svd(Env, svd_linds, mindim=chi, maxdim=chi)
+                    # update guess at row
+                    guess[row]    = tsvd.U
+                    guess[row+1] *= tsvd.S*tsvd.V
+                elseif !isodd(sweep) && row > 1
+                    svd_linds     = setdiff(inds(guess[row]),IndexSet(commonind(guess[row], guess[row-1])))
+                    tsvd          = svd(Env, svd_linds, mindim=chi, maxdim=chi)
+                    # update guess at row
+                    guess[row]    = tsvd.U
+                    guess[row-1] *= tsvd.S*tsvd.V
+                else
+                    # update guess at row
+                    guess[row]   = copy(Env) 
+                end
+                Env_below    *= guess[row]
+            end
+        end
+    end
+    return guess
+end
+
+# now with two-site
+function fitPEPSMPO(A::fPEPS, prev_mps::Vector{<:ITensor}, ops::Vector{ITensor}, col::Int, chi::Int)
+    Ny, Nx = size(A)
+    is_cu  = is_gpu(A)
+    @timeit "make guess" begin
+        guess::MPS = initGuessRandomMPS(A, prev_mps, col, chi)
+        if is_cu
+            guess = cuMPS(guess)
+        end
+    end
+    for sweep in 1:1
+        @timeit "make Envs_above" begin
+            Envs_above = constructEnvsAbove(A, guess, prev_mps, ops, col, sweep)
+        end
+        order     = isodd(sweep) ? (1:Ny-1) : reverse(1:Ny-1)
+        Env_below = is_cu ? cuITensor(1.0) : ITensor(1.0)
+        @inbounds for row in order
+            # construct environment for the row, have to do this every time
+            @timeit "build row Env" begin
+                Env  = copy(Env_below)
+                for row_ in (row, row+1)
+                    tmp = A[row_, col]
+                    tmp *= ops[row_]
+                    tmp *= prev_mps[row_]
+                    Env *= tmp
+                    Env *= dag(prime(A[row_, col]))
```

Comment:
I believe this is a D^6. However, if the guess MPS can be used here, the algorithm can be reduced to a D^5 until one builds the part with the two guess-MPS tensors missing. Is that this part?

Reply:
This is the part with the two tensors missing, yes